ethereum/ethevents/events.go

```diff
@@ -172,155 +161,81 @@ func (ev *EthereumEvents) AddEventHandler(handler EventHandler) {
-	var sub eth.Subscription
-	var err error
+func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context) {
+
 	// Get current block
 	tctx, cancel := context.WithTimeout(ctx, types.EthereumReadTimeout)
 	defer cancel()
-	var lastBlock int64
-	var blk *ethtypes.Block
-	errors := 0
-	for {
-		if errors > 5 {
-			log.Fatal("the web3 client connection is not working")
-		}
-		blk, err = ev.VotingHandle.EthereumClient.BlockByNumber(tctx, nil)
-		if err != nil {
-			log.Errorf("cannot get ethereum block: %s", err)
-			errors++
-			time.Sleep(time.Second * 2)
-			if err := ev.VotingHandle.Connect(ev.DialAddrs); err != nil {
-				log.Warn(err)
-			}
-			continue
-		}
-		lastBlock = blk.Number().Int64()
-		break
+	var err error
+	uLastBlockNumber, err := ev.VotingHandle.EthereumClient.BlockNumber(tctx)
+	if err != nil {
+		log.Fatalf("cannot get last block number: (%v)", err)
 	}
-
-	if fromBlock != 0 {
-		// do not retry if error processing old logs
-		// Unless this is the first set of oracles, it is almost
-		// sure that the events are already processed so do not
-		// block and do not fatal here
-		if err := ev.processEventLogsFromTo(ctx, fromBlock, lastBlock,
-			ev.VotingHandle.EthereumClient); err != nil {
-			log.Errorf("cannot process event logs: %v", err)
-		}
-		// Update block number
-		// Expect the client to be connected here
-		// and the call is successful, if not
-		// block the execution and alert
-		errors = 0
-		for {
-			if errors > 5 {
-				log.Fatal("the web3 connection is not working")
-			}
-			tctx, cancel := context.WithTimeout(ctx, types.EthereumReadTimeout)
-			defer cancel()
-			blk, err = ev.VotingHandle.EthereumClient.BlockByNumber(tctx, nil)
-			if err != nil {
-				log.Errorf("cannot update block number: %v", err)
-				errors++
-				time.Sleep(time.Second * 2)
-				// If any error try close the client and reconnect
-				if err := ev.VotingHandle.Connect(ev.DialAddrs); err != nil {
-					log.Warn(err)
-				}
-				continue
-			}
-			lastBlock = blk.Number().Int64()
-			break
-		}
-		// For security, read also the new passed blocks before subscribing
-		// Same as the last processEventLogsFromTo function call
-		if err := ev.processEventLogsFromTo(ctx, lastBlock, blk.Number().Int64(),
-			ev.VotingHandle.EthereumClient); err != nil {
-			log.Errorf("cannot process event logs: %s", err)
-		}
-	} else {
-		// For security, even if subscribe only, force to process at least the past `readBlockPast`
-		// Same as the last processEventLogsFromTo function call
-		if err := ev.processEventLogsFromTo(ctx, blk.Number().Int64()-readBlocksPast,
-			blk.Number().Int64(), ev.VotingHandle.EthereumClient); err != nil {
-			log.Errorf("cannot process event logs: %v", err)
-		}
+	lastBlockNumber := int64(uLastBlockNumber)
+	fromBlock := ev.EthereumLastKnownBlock.Load().(int64)
+	if fromBlock == 0 {
+		ev.EthereumLastKnownBlock.Store(lastBlockNumber)
```

Comment:
note that this Store is separate from the Load, so it's not atomic. I imagine we don't mind that. If the original atomic value was 0 and two goroutines both try to update it at the same time, I imagine they'd both try to update it to the same new value. If it's possible that those two goroutines would end up with different "latest block" values, then that might be problematic.